src/Suave/Logging/LogLine.fs

```diff
@@ -13,23 +15,45 @@ type LogLine =
     level         : LogLevel
     /// the source of the log line, e.g. 'ModuleName.FunctionName'
     path          : string
-    /// the message that the application wants to log
-    message       : string
+    /// the template message that the application wants to log
+    template      : string
+    /// a function which can format the log template message. When the
+    /// message is pre-formatted, this is None. When the log message 
+    /// contains structured data, this function can be used to render
+    /// the message template with the data.
+    formatter     : LogTemplateFormatter option
+    /// any key-value based data to log
+    data          : Map<string, obj>
     /// an optional exception
     ``exception`` : exn option
     /// timestamp when this log line was created
     tsUTCTicks    : int64 }
-    
+
+    /// Formats the log line as a string
+    member x.formatMessage () =
+        match x.formatter with
+        | None -> x.template
+        | Some fmter -> fmter x.template x.GetDataByName
+
+    /// Gets a data value
+    member x.GetDataByName name = Option.toObj (x.data.TryFind name)
+
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module LogLine =
   let mk path level trace ex message =
-    { message       = message
+    { template      = message
+      formatter     = None
+      data          = Map.empty
       level         = level
       path          = path
       ``exception`` = ex
       trace         = trace
       tsUTCTicks  = Globals.utcNow().Ticks }
 
+  /// assists with type inference, avoiding ugly 'box' calls for each
+  /// value item when constructing the log data map 
+  let mapLogData (data: (string*obj) list) = Map.ofList data
```

Comment:
I think we are getting somewhere too :)

Parsing the template before it leaves the call site has benefits:
 1. Ability to match message fields to their names (positionally), without being required to specify the names twice.
 2. Ability to warn the message author earlier if there is a problem.

The downsides I can think of:
 1. overhead of parsing OR complexity of caching them. This is a fairly small overhead, and it's already been fairly well optimised with lots of ugly code and tests. Serilog, for example, actually caches the parsed template messages to avoid this overhead. You would need to be doing many thousands of log messages before it becomes an issue worth considering (which seems quite possible with Suave and Logary and certainly MessageTemplates in general). If parsed templates were cached, we have the added complexity, but we can learn from the early mistakes [Serilog made](https://github.com/serilog/serilog/issues/523).